### PR TITLE
IA-547 modules federation

### PR DESCRIPTION
**Pull request by beygorghor**
Screenshot 2021-05-19 at 23 25 43
Screenshot 2021-05-19 at 23 25 36

DONE:
- removed 6 unused packages
- Upgrade to Node 15.14, NPM 7, WEBPACK 5 and Node sass 5
- split package json in two and add workspace in the root one
-  move eslint, prettier, test and co to application folder + fix import issues
- adapt webpack dev and prod to allow a plugin with ModuleFederationPlugin
- change docker config to copy the new package json
- create a standalone react app using webpack 5 as plugin sample => https://github.com/BLSQ/iaso-sample-plugin
- fetch plugin in iaso root applilcation, adapt the routes and the sidebar menu with the plugin entries
- fail silently if plugin not found or not in the config

Don't forget to remove docker iaso images and container (except DB) to make it work

Don not install packages in the iaso-root folder, module federation will take care of this.

To launch the tests, remove node-modules from the root of the app and run `npm intall ` then `npm run test` or `npm run test-coverage`

TODO:
- find a proper way to include the plugin(s) into `webpack.dev.js` and `webpack.prod.js` (from python settings)
- Use pythoon settings to display or not link to `http://localhost:3001/remoteEntry.js` in base.htrml template
- maybe fix github action
- pass feature flags, mui theme, active locale to children and manage translations

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -21,10 +21,10 @@ jobs:
     steps:
       - uses: actions/checkout@v2
 
-      - name: Use node.js 10
+      - name: Use node.js 15
```
Comment (geoffreydhuyvetters) [inline .github/workflows/main.yml:24]: Specific reason we're using 15 here? LTS is 14

Comment (beygorghor) [reply .github/workflows/main.yml:24]: Yes, 15 is required but don't remember why...

Comment (beygorghor) [reply .github/workflows/main.yml:24]: Now that it's working we can try with 14 in fact

**Comment by quang-le** [comment]
> If we are going to reorganise everything, maybe we could move the JS app to a top dir? Not sure such a path still make sense hat/assets/js/apps/Iaso/domains/app/components/MenuItemComponent.js or hat/assets/js/apps/Iaso/css/_leafletCluster.scss?

True, although I'd say it's out of scope for this PR, as I reckon it would require checking if nothing breaks in Django

File: hat/assets/js/apps/Iaso/webpack.dev.js [hunk 2]
```diff
@@ -161,3 +157,21 @@ module.exports = {
...
+    new webpack.DefinePlugin({
+        'process.env.PLUGIN_1': JSON.stringify('test_app/pluginConfig'),
+    }),
+    new ModuleFederationPlugin({
+        name: 'iaso_root',
+        library: { type: 'var', name: 'iaso_root' },
+        remotes: {
+            test_app: 'test_app',
+        },
+    }),
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/webpack.dev.js:174]: We also need to investigate how to properly use the shared dependencies that we can pass to the `ModuleFederationPlugin`

File: hat/templates/iaso/base.html [hunk 3]
```diff
@@ -39,6 +39,7 @@
   <meta property="og:image" content="" />
   <meta property="og:url" content="http://iaso.bluesquarehub.com" />
   {% endif %}
+  <script src="http://localhost:3001/remoteEntry.js"></script>
```
Comment (olethanh) [inline hat/templates/iaso/base.html:42]: I'm a bit curious what is this, and how will it works in prod? the filename isn't  very reassuring either

File: hat/dashboard/urls.py [hunk 4]
```diff
@@ -13,6 +13,7 @@
     url(r"^orgunits/groups.*$", views.iaso, name="iaso_orgunits_groups"),
     url(r"^orgunits/types.*$", views.iaso, name="iaso_orgunits_types"),
     url(r"^completeness.*$", views.iaso, name="iaso_completeness"),
+    url(r"^plugin.*$", views.iaso, name="iaso_plugin"),
```
[the remark this answers is not in the document]
Comment (beygorghor) [reply hat/dashboard/urls.py:16]: We need to find a way to provide plugin config with settings.py and docker.yml, this url has been added for test purpose. The main concern is more how to pass plugin config to webpack.dev an webpack.prod

File: hat/templates/iaso/base.html (hunk 3 above)
Comment (beygorghor) [reply hat/templates/iaso/base.html:42]: again this need to be display only id dev mode and if a plugin config is provided (this link should be in the config)

File: hat/dashboard/urls.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (olethanh) [reply hat/dashboard/urls.py:16]: sorry I misread it's just plugin and not polio, should not be a problem then to left it as is

**Comment by beygorghor** [comment]
So the docker file should contain the configs of the plugin(s) and pass it to the settings.py.
A plugin config needs to contain:
- the plugin name e.g:"test_app" (used in webpack and in iaso main app to fetch plugin config)
- the path to use in dev mode to connect the plugin to Iaso   e.g:"http://localhost:3001/remoteEntry.js" (used in base.html in dev mode)

After that we need to find a way to pass this config to webpack.dev.js and webpack.prod.js and generate env variables with the plugin config name so we can import this config into our main app.
Regarding this, we have this issue => https://github.com/webpack/webpack/issues/12167#issuecomment-742644667
@olethanh if you want to have a look, you are welcome.

**Comment by olethanh** [comment]
Just to be sure , does this mean that in dev mode there is going to be two webpack servers the current one at port 3000 and a new one at port 3001? And for each new plugin we will have another one?

**Pull request closed without merging**